### PR TITLE
AzureEnvironment as part of the AzureClusterSpec instead of an environment variable 

### LINKED ISSUE #1227: Move AZURE_ENVIRONMENT to AzureCluster spec
/kind feature

**Describe the solution you'd like**
[A clear and concise description of what you want to happen.]
Moving AZURE_ENVIRONMENT to AzureCluster spec will allow capz `infrastructure-components.yaml` to be deployed without requiring `envsubst` to set a default value

**Anything else you would like to add:**
[Miscellaneous information that will assist in solving the issue.]
source: https://github.com/kubernetes-sigs/cluster-api/issues/3881#issuecomment-743370519 

**Environment:**

- cluster-api-provider-azure version: 
- Kubernetes version: (use `kubectl version`): 
- OS (e.g. from `/etc/os-release`):

### PR DESCRIPTION
**Pull request by nader-ziada**
**What type of PR is this?**
/kind feature
/kind api-change

**What this PR does / why we need it**:
-  Moving AZURE_ENVIRONMENT to AzureCluster spec will allow capz infrastructure-components.yaml to be deployed without requiring envsubst to set a default value

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #1227

**Special notes for your reviewer**:

_Please confirm that if this PR changes any image versions, then that's the sole change this PR makes._

**TODOs**:

- [ ] squashed commits
- [ ] includes documentation
- [ ] adds unit tests

**Release note**:

```release-note
Set AzureEnvironment as part of the AzureClusterSpec instead of an environment variable. This is a BREAKING CHANGE to users not using AzurePublicCloud.
action required: Users will need to update the AzureCluster resource to include the appropriate value of AzureEnvironment 
```

### CODE REVIEW

**Review by devigned — CHANGES_REQUESTED** [review]
Nice!

One thing that would be good to add here is a update hook to reject changes to the `AzureEnvironment`. To change this field would imply a delete of the cluster and rebuilding the cluster in another Azure cloud. I don't think we want to enable that.

**Comment by nader-ziada** [comment]
> One thing that would be good to add here is a update hook to reject changes to the `AzureEnvironment`. To change this field would imply a delete of the cluster and rebuilding the cluster in another Azure cloud. I don't think we want to enable that.

good catch, didn't think about that, but makes perfect sense

**Review by devigned — APPROVED** [review]
/lgtm
/assign @CecileRobertMichon

**Comment by nader-ziada** [comment]
oh wait, forgot to update relevant docs as well
/hold

**Review by devigned — APPROVED** [review]
/lgtm

File: config/default/manager_cloud_env_patch.yaml [hunk 1]
```diff
@@ -1,16 +0,0 @@
...
-metadata:
-  name: controller-manager
-  namespace: system
-spec:
-  template:
-    spec:
-      containers:
-      - name: manager
-        env:
-          - name: AZURE_ENVIRONMENT
```
Comment (CecileRobertMichon) [inline config/default/manager_cloud_env_patch.yaml:12]: By removing this we're essentially making it not supported to set the AZURE_ENVIRONMENT value via en env variable right? Wasn't the idea to keep it as optional for back compat?

Comment (nader-ziada) [reply config/default/manager_cloud_env_patch.yaml:12]: If we keep it we would still need to use envsubst with the manifest, right?

Comment (CecileRobertMichon) [reply config/default/manager_cloud_env_patch.yaml:12]: Ok but this is a breaking change then for users not using AzurePublicCloud, let's make sure the release note calls that out

Comment (nader-ziada) [reply config/default/manager_cloud_env_patch.yaml:12]: should we leave this in and just not set a default for it using envsubst? would that be easier for not breaking users

Comment (CecileRobertMichon) [reply config/default/manager_cloud_env_patch.yaml:12]: > should we leave this in and just not set a default for it using envsubst? would that be easier for not breaking users

that would make it a required variable with clusterctl which might impact even more users (ie. everyone in AzurePublicCloud relying on the default, like quickstart users).

I think removing it and adding a clear release note for non-public cloud users to set the value is the best course of action.

**Pull request merged**